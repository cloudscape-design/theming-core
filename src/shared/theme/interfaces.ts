// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
export interface OptionalState {
  selector: string;
  media?: string;
}

export interface DefaultState {
  default: boolean;
}
export interface Mode {
  id: string;
  states: Record<string, DefaultState | OptionalState>;
}

export type Token = string;
export type Value = string;
export type Reference = string;

export type GlobalValue = Value;
export type GlobalReference = Reference;
export type ModeValue<S extends string = string> = Record<S, Value | Reference>;
export type Assignment = GlobalValue | GlobalReference | ModeValue;

export interface Context {
  id: string;
  selector: string;
  tokens: Record<string, Assignment>;
}

export interface Theme {
  id: string;
  selector: string;
  tokens: Record<string, Assignment>;
  modes: Record<string, Mode>;
  tokenModeMap: Record<string, string>;
  contexts: Record<string, Context>;
  referenceTokens?: ReferenceTokens;
}

/**
 * Reference tokens enable seed-based palette generation and semantic token organization.
 */
export interface ReferenceTokens {
  color?: ColorReferenceTokens;
}

export interface ColorReferenceTokens {
  primary?: ColorPaletteInput;
  neutral?: ColorPaletteInput;
  error?: ColorPaletteInput;
  success?: ColorPaletteInput;
  warning?: ColorPaletteInput;
  info?: ColorPaletteInput;
}

/**
 * Color reference tokens organized by semantic color categories.
 * Each category can be defined using a seed color (string) or detailed palette definition.
 */
export type ColorPaletteInput = string | ColorPaletteDefinition;

/**
 * Palette steps available across all color types. Different color categories
 * may use different subsets of these steps.
 */
export type PaletteStep =
  | 50
  | 100
  | 150
  | 200
  | 250
  | 300
  | 350
  | 400
  | 450
  | 500
  | 550
  | 600
  | 650
  | 700
  | 750
  | 800
  | 850
  | 900
  | 950
  | 1000;

/**
 * Color palette definition supporting both seed generation and explicit overrides.
 */
export interface ColorPaletteDefinition extends Partial<Record<PaletteStep, string>> {
  seed?: string;
}

type Tokens = Partial<Record<string, GlobalValue | TypedModeValueOverride>>;

export interface Override {
  tokens: Tokens;
  contexts?: Record<string, { tokens: Tokens } | undefined>;
}

export type TypedModeValueOverride<S extends string = string> = Partial<Record<S, Value>>;
/**
 * A theme preset contains information to render themed components or applying
 * a theme at runtime. It should only be used internally and not exposed to customers.
 */
export interface ThemePreset {
  theme: Theme;
  secondary?: Theme[];
  themeable: Token[];
  /** List of public design tokens */
  exposed: Token[];
  /** Map between design tokens and custom properties */
  propertiesMap: Record<Token, string>;
  /** Map between design tokens and variable names */
  variablesMap: Record<Token, string>;
}
