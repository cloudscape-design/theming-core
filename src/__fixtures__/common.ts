// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import {
  Theme,
  Mode,
  Context,
  Override,
  ThemePreset,
  FullResolution,
  SpecificResolution,
  FullResolutionPaths,
} from '../shared/theme';

export function createStubPropertiesMap(theme: Theme): Record<string, string> {
  return Object.keys(theme.tokens).reduce((acc, curr) => {
    acc[curr] = `--${curr}-css`;
    return acc;
  }, {} as Record<string, string>);
}

export function createStubVariablesMap(theme: Theme): Record<string, string> {
  return Object.keys(theme.tokens).reduce((acc, curr) => {
    acc[curr] = `${curr}-var`;
    return acc;
  }, {} as Record<string, string>);
}

export const colorMode: Mode = {
  id: 'color',
  states: {
    light: { default: true },
    dark: { selector: '.dark', media: 'not print' },
  },
};

export const densityMode: Mode = {
  id: 'density',
  states: {
    compact: { selector: '.compact' },
    comfortable: { default: true },
  },
};

export const motionMode: Mode = {
  id: 'motion',
  states: {
    default: { default: true },
    disabled: { selector: '.disabled-motion' },
  },
};

export const navigationContext: Context = {
  id: 'navigation',
  selector: '.navigation',
  tokens: {
    shadow: {
      light: '{black}', // example of a reference
      dark: '{brown}',
    },
    boxShadow: {
      dark: 'purple', // example of a literal value
      light: 'purple',
    },
  },
};

const anotherContext: Context = {
  id: 'another-context',
  selector: '.another-context',
  tokens: {
    shadow: {
      light: '{black}', // example of a reference
      dark: '{brown}',
    },
    boxShadow: {
      dark: 'purple', // example of a literal value
      light: 'purple',
    },
  },
};

const emptyTheme: Theme = {
  id: 'root',
  selector: ':root',
  tokens: {},
  tokenModeMap: {},
  contexts: {},
  modes: {},
};

export const rootTheme: Theme = {
  id: 'root',
  selector: ':root',
  tokens: {
    fontFamilyBase: '"Helvetica Neue", Arial, sans-serif',
    fontFamilyBody: '{fontFamilyBase}',
    black: 'black',
    grey: 'grey',
    brown: 'brown',
    shadow: {
      light: '{grey}',
      dark: '{black}',
    },
    buttonShadow: `{shadow}`,
    boxShadow: {
      light: '{shadow}',
      dark: '{brown}',
    },
    lineShadow: {
      light: '{buttonShadow}',
      dark: '{boxShadow}',
    },
    small: '1px',
    medium: '3px',
    scaledSize: {
      compact: '{small}',
      comfortable: '{medium}',
    },
    appear: {
      default: '20ms',
      disabled: '0',
    },
    containerShadowBase: '2px 3px orange, -1px 0 8px olive',
    modalShadowContainer: '{containerShadowBase}',
  },
  tokenModeMap: {
    shadow: 'color',
    buttonShadow: 'color',
    boxShadow: 'color',
    lineShadow: 'color',
    scaledSize: 'density',
    appear: 'motion',
  },
  contexts: {
    navigation: navigationContext,
  },
  modes: {
    density: densityMode,
    color: colorMode,
    motion: motionMode,
  },
};

export const secondaryTheme: Theme = {
  ...rootTheme,
  id: 'secondary',
  selector: '.secondary-theme',
  tokens: {
    ...rootTheme.tokens,
    black: 'purple',
    brown: 'black',
  },
  contexts: {
    navigation: {
      ...navigationContext,
      tokens: {
        shadow: {
          light: '{black}',
          dark: '{grey}',
        },
      },
    },
  },
};

const anotherSecondaryTheme: Theme = {
  ...secondaryTheme,
  contexts: {
    navigation: {
      ...navigationContext,
    },
    'another-context': {
      ...anotherContext,
    },
  },
};

export const themesWithCircularDependencies: Array<Theme> = [
  {
    ...emptyTheme,
    tokens: {
      firstToken: '{firstToken}',
    },
  },
  {
    ...emptyTheme,
    tokens: {
      firstToken: '{secondToken}',
      secondToken: '{thirdToken}',
      thirdToken: '{secondToken}',
    },
  },
];

export const themeWithNonExistingToken: Theme = {
  ...emptyTheme,
  tokens: {
    firstToken: '{nonExistingToken}',
  },
};

export const fullResolution: FullResolution = {
  black: 'black',
  grey: 'grey',
  brown: 'brown',
  shadow: {
    light: 'grey',
    dark: 'black',
  },
  buttonShadow: {
    light: 'grey',
    dark: 'black',
  },
  boxShadow: {
    light: 'grey',
    dark: 'brown',
  },
  lineShadow: {
    light: 'grey',
    dark: 'brown',
  },
  small: '1px',
  medium: '3px',
  scaledSize: {
    compact: '1px',
    comfortable: '3px',
  },
  appear: {
    default: '20ms',
    disabled: '0',
  },
  containerShadowBase: '2px 3px orange, -1px 0 8px olive',
  modalShadowContainer: '2px 3px orange, -1px 0 8px olive',
  fontFamilyBase: '"Helvetica Neue", Arial, sans-serif',
  fontFamilyBody: '"Helvetica Neue", Arial, sans-serif',
};

export const fullResolutionPaths: FullResolutionPaths = {
  black: ['black'],
  grey: ['grey'],
  brown: ['brown'],
  shadow: {
    light: ['shadow', 'grey'],
    dark: ['shadow', 'black'],
  },
  buttonShadow: {
    light: ['buttonShadow', 'shadow', 'grey'],
    dark: ['buttonShadow', 'shadow', 'black'],
  },
  boxShadow: {
    light: ['boxShadow', 'shadow', 'grey'],
    dark: ['boxShadow', 'brown'],
  },
  lineShadow: {
    light: ['lineShadow', 'buttonShadow', 'shadow', 'grey'],
    dark: ['lineShadow', 'boxShadow', 'brown'],
  },
  small: ['small'],
  medium: ['medium'],
  scaledSize: {
    compact: ['scaledSize', 'small'],
    comfortable: ['scaledSize', 'medium'],
  },
  appear: {
    default: ['appear'],
    disabled: ['appear'],
  },
  containerShadowBase: ['containerShadowBase'],
  modalShadowContainer: ['modalShadowContainer', 'containerShadowBase'],
  fontFamilyBase: ['fontFamilyBase'],
  fontFamilyBody: ['fontFamilyBody', 'fontFamilyBase'],
};

export const defaultsResolution: SpecificResolution = {
  black: 'black',
  grey: 'grey',
  brown: 'brown',
  shadow: 'grey',
  buttonShadow: 'grey',
  boxShadow: 'grey',
  lineShadow: 'grey',
  small: '1px',
  medium: '3px',
  scaledSize: '3px',
  appear: '20ms',
  fontFamilyBase: '"Helvetica Neue", Arial, sans-serif',
  fontFamilyBody: '"Helvetica Neue", Arial, sans-serif',
  containerShadowBase: '2px 3px orange, -1px 0 8px olive',
  modalShadowContainer: '2px 3px orange, -1px 0 8px olive',
};

export const override: Override = {
  tokens: {
    shadow: {
      light: 'yellow',
      dark: 'orange',
    },
    boxShadow: {
      light: 'green',
    },
    lineShadow: {
      light: 'pink',
      dark: 'pink',
    },
    buttonShadow: 'red',
  },
  contexts: {
    navigation: {
      tokens: {
        shadow: {
          light: 'pink',
        },
        buttonShadow: {
          dark: 'green',
        },
      },
    },
  },
};

export const overrideWithRandomContext: Override = {
  tokens: { ...override.tokens },
  contexts: {
    randomContext: {
      tokens: {
        shadow: {
          light: 'pink',
        },
        buttonShadow: {
          dark: 'green',
        },
      },
    },
  },
};

export const preset: ThemePreset = {
  theme: rootTheme,
  themeable: ['shadow', 'boxShadow', 'buttonShadow', 'lineShadow'],
  exposed: ['shadow', 'buttonShadow', 'boxShadow', 'lineShadow'],
  propertiesMap: createStubPropertiesMap(rootTheme),
  variablesMap: createStubVariablesMap(rootTheme),
};

export const presetWithSecondaryTheme: ThemePreset = {
  ...preset,
  secondary: [secondaryTheme],
};
export const anotherPresetWithSecondaryTheme: ThemePreset = {
  ...preset,
  secondary: [anotherSecondaryTheme],
};

export const descriptions: Record<string, string> = {
  shadow: 'shadow',
  buttonShadow: 'button shadow',
  boxShadow: 'box shadow',
  lineShadow: 'line shadow',
};

const themeWithValidSchema: Theme = {
  id: 'valid',
  selector: ':root',
  tokens: {
    fontFamilyBase: '"Helvetica Neue", Arial, sans-serif',
    colorButton: '#ffffff',
    shadowContainer: 'rgba(0, 7, 22, 0.2)',
    spaceLarge: '26px',
    borderRadiusAlert: '4px',
    motionDurationOne: '5s',
    motionEasingTwo: 'whatever',
    motionKeyframesThree: 'something',
  },
  tokenModeMap: {
    colorButton: 'color',
    spaceLarge: 'density',
    shadowContainer: 'color',
    motionDurationOne: 'motion',
    motionEasingTwo: 'motion',
    motionKeyframesThree: 'motion',
  },
  contexts: {
    navigation: {
      id: 'navigation',
      selector: '.navigation',
      tokens: {
        colorButton: '#000000',
      },
    },
  },
  modes: {
    density: densityMode,
    color: colorMode,
    motion: motionMode,
  },
};

const exposed = [
  'fontFamilyBase',
  'colorButton',
  'shadowContainer',
  'spaceLarge',
  'borderRadiusAlert',
  'motionDurationOne',
  'motionEasingTwo',
  'motionKeyframesThree',
];

export const presetWithValidSchema: ThemePreset = {
  theme: themeWithValidSchema,
  themeable: exposed,
  exposed: exposed,
  propertiesMap: createStubPropertiesMap(themeWithValidSchema),
  variablesMap: {
    fontFamilyBase: 'font-family-base',
    colorButton: 'color-button',
    shadowContainer: 'shadow-container',
    spaceLarge: 'space-large',
    borderRadiusAlert: 'border-radius-alert',
    motionDurationOne: 'motion-duration-one',
    motionEasingTwo: 'motion-easing-two',
    motionKeyframesThree: 'motion-keyframes-three',
  },
};

export const descriptionsForValidTheme = {
  fontFamilyBase: 'font family',
  colorButton: 'color',
  shadowContainer: 'shadow',
};
