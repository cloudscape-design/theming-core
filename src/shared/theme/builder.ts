// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import {
  Theme,
  Context,
  Mode,
  ReferenceTokens,
  ColorReferenceTokens,
  ColorPaletteInput,
  PaletteStep,
} from './interfaces';

export type TokenCategory<T extends string, V> = Record<T, V>;

export class ThemeBuilder {
  theme: Theme;

  constructor(id: string, selector: string, modes: Mode[]) {
    this.theme = {
      id,
      selector,
      modes: modes.reduce((acc, curr) => {
        acc[curr.id] = curr;
        return acc;
      }, {} as Record<string, Mode>),
      tokens: {},
      contexts: {},
      tokenModeMap: {},
    };
  }

  addTokens<T extends string, V>(tokens: TokenCategory<T, V>, mode?: Mode): ThemeBuilder {
    this.theme.tokens = {
      ...this.theme.tokens,
      ...tokens,
    };

    if (mode) {
      const modes = Object.keys(tokens).reduce((acc, token) => {
        acc[token] = mode.id;
        return acc;
      }, {} as Record<string, string>);
      this.theme.tokenModeMap = {
        ...this.theme.tokenModeMap,
        ...modes,
      };
    }

    return this;
  }

  addContext(context: Context): ThemeBuilder {
    this.theme.contexts[context.id] = context;
    return this;
  }

  addReferenceTokens(referenceTokens: ReferenceTokens): ThemeBuilder {
    this.theme.referenceTokens = referenceTokens;

    // Process reference tokens and add generated tokens to theme
    if (referenceTokens.color) {
      const generatedTokens = this.processReferenceTokens(referenceTokens.color);
      this.theme.tokens = { ...this.theme.tokens, ...generatedTokens };
    }

    return this;
  }

  build(): Theme {
    return this.theme;
  }

  private processReferenceTokens(colorTokens: ColorReferenceTokens): Record<string, string> {
    const generatedTokens: Record<string, string> = {};

    Object.entries(colorTokens).forEach(([colorName, paletteInput]) => {
      const palette = this.processColorPaletteInput(paletteInput);

      // Add generated palette tokens with naming convention: colorPrimary50, colorPrimary600, etc.
      Object.entries(palette).forEach(([step, value]) => {
        const tokenName = `color${this.capitalize(colorName)}${step}`;
        generatedTokens[tokenName] = value;
      });
    });

    return generatedTokens;
  }

  private processColorPaletteInput(input: ColorPaletteInput): Record<PaletteStep, string> {
    if (typeof input === 'string') {
      // Simple seed case: generate basic palette from seed
      return this.generateBasicPalette(input);
    } else {
      // Complex case: object with seed and/or explicit step values
      const validSteps = [
        50, 100, 150, 200, 250, 300, 350, 400, 450, 500, 550, 600, 650, 700, 750, 800, 850, 900, 950, 1000,
      ];

      // Start with generated palette if seed is provided, otherwise empty
      const result: Record<PaletteStep, string> = input.seed
        ? this.generateBasicPalette(input.seed)
        : ({} as Record<PaletteStep, string>);

      // Override with explicit step values
      Object.entries(input).forEach(([step, value]) => {
        const numStep = Number(step);
        if (step !== 'seed' && value && validSteps.includes(numStep)) {
          result[numStep as PaletteStep] = value;
        }
      });

      return result;
    }
  }

  private generateBasicPalette(seed: string): Record<PaletteStep, string> {
    // Placeholder implementation - will be replaced with HCT generation later
    // For now, just return the seed color for all steps to establish the plumbing
    return {
      50: seed,
      100: seed,
      150: seed,
      200: seed,
      250: seed,
      300: seed,
      350: seed,
      400: seed,
      450: seed,
      500: seed,
      550: seed,
      600: seed,
      650: seed,
      700: seed,
      750: seed,
      800: seed,
      850: seed,
      900: seed,
      950: seed,
      1000: seed,
    };
  }

  private capitalize(str: string): string {
    return str.charAt(0).toUpperCase() + str.slice(1);
  }
}
