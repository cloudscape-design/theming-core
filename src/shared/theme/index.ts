// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
export {
  Theme,
  ThemePreset,
  Override,
  Mode,
  Context,
  Value,
  OptionalState,
  GlobalValue,
  ModeValue,
  TypedModeValueOverride,
} from './interfaces';
export { ThemeBuilder, TokenCategory } from './builder';
export {
  resolveTheme,
  resolveThemeWithPaths,
  resolveContext,
  reduce,
  defaultsReducer,
  modeReducer,
  difference,
  FullResolution,
  SpecificResolution,
  FullResolutionPaths,
} from './resolve';
export { validateOverride } from './validate';
export { merge, mergeInPlace } from './merge';
