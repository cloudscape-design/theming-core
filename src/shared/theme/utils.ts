// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { Assignment, DefaultState, OptionalState, Theme } from './interfaces';
import { Value, Reference, ModeValue, Mode } from './interfaces';

export function isValue(val: unknown): val is Value {
  return typeof val === 'string' && !isReference(val);
}

export function isReference(val: unknown): val is Reference {
  return typeof val === 'string' && val.charAt(0) === '{' && val.charAt(val.length - 1) === '}';
}

export function isModeValue(val: unknown): val is ModeValue {
  return (
    typeof val === 'object' &&
    val !== null &&
    !Object.keys(val).some((state) => !(isValue((val as ModeValue)[state]) || isReference((val as ModeValue)[state])))
  );
}

export function areAssignmentsEqual(valueA: Assignment, valueB: Assignment): boolean {
  return (
    valueA === valueB ||
    (typeof valueA === 'object' &&
      typeof valueB == 'object' &&
      Object.keys(valueA).length === Object.keys(valueB).length &&
      Object.keys(valueA).every((key) => valueA[key] === valueB[key]))
  );
}

export function isOptionalState(val: DefaultState | OptionalState): val is OptionalState {
  return 'selector' in val;
}

export function getReference(reference: Reference): string {
  return reference.slice(1, reference.length - 1);
}

export function getMode(theme: Theme, token: string): Mode | null {
  const modeId = theme.tokenModeMap[token];
  return theme.modes[modeId] ?? null;
}

export function getDefaultState(mode: Mode): string {
  const states = Object.keys(mode.states);
  for (let index = 0; index < states.length; index++) {
    const state = states[index];
    const option = mode.states[state];
    if (option && 'default' in option && option.default) {
      return state;
    }
  }
  throw new Error(`Mode ${JSON.stringify(mode)} does not have a default state`);
}
