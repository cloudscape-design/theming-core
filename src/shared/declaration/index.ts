// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { mergeInPlace, Override, Theme } from '../theme';
import type { PropertiesMap, SelectorCustomizer } from './interfaces';
import { RuleCreator } from './rule';
import { SingleThemeCreator } from './single';
import { MultiThemeCreator } from './multi';
import { Selector } from './selector';
import { AllPropertyRegistry, UsedPropertyRegistry } from './registry';
import { MinimalTransformer } from './transformer';
import { cloneDeep, values } from '../utils';

function createMinimalTheme(base: Theme, override: Override): Theme {
  const minimalTheme = cloneDeep(base);
  const contextTokens: Set<string> = new Set();
  values(minimalTheme.contexts).forEach((context) => {
    Object.keys(context.tokens).forEach((key) => {
      if (!(key in override.tokens) && !(key in (override?.contexts?.[context.id]?.tokens ?? {}))) {
        delete context.tokens[key];
      } else {
        contextTokens.add(key);
      }
    });
  });
  Object.keys(minimalTheme.tokens).forEach((key) => {
    if (!contextTokens.has(key) && !(key in override.tokens)) {
      delete minimalTheme.tokens[key];
    }
  });
  return mergeInPlace(minimalTheme, override);
}

export function createOverrideDeclarations(
  base: Theme,
  override: Override,
  propertiesMap: PropertiesMap,
  selectorCustomizer: SelectorCustomizer
): string {
  // create theme containing only modified tokens
  const minimalTheme = createMinimalTheme(base, override);
  const ruleCreator = new RuleCreator(new Selector(selectorCustomizer), new AllPropertyRegistry(propertiesMap));
  const stylesheetCreator = new SingleThemeCreator(minimalTheme, ruleCreator, base);
  const stylesheet = stylesheetCreator.create();
  return stylesheet.toString();
}

export function createBuildDeclarations(
  primary: Theme,
  secondary: Theme[],
  propertiesMap: PropertiesMap,
  selectorCustomizer: SelectorCustomizer,
  used: string[]
): string {
  const ruleCreator = new RuleCreator(new Selector(selectorCustomizer), new UsedPropertyRegistry(propertiesMap, used));
  const stylesheetCreator = new MultiThemeCreator([primary, ...secondary], ruleCreator);
  const stylesheet = stylesheetCreator.create();
  const transformer = new MinimalTransformer();
  const minimal = transformer.transform(stylesheet);
  return minimal.toString();
}
