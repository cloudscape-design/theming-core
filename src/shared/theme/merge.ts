// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { Theme, Override, Assignment } from './interfaces';
import { cloneDeep, entries } from '../utils';
import { getMode, isModeValue, isReference, isValue } from './utils';

/**
 * This function applies all tokens from the override to the theme.
 * It returns the resulting theme. The original theme object is not
 * modified.
 */
export function merge(theme: Theme, override: Override): Theme {
  const result = cloneDeep(theme);

  function withTokenApplied(
    originalValue: Assignment,
    token: string,
    update: (typeof override.tokens)[string]
  ): Assignment | undefined {
    const isGlobal = isValue(update) || isReference(update);
    const mode = getMode(theme, token);

    if (mode && isGlobal) {
      return Object.keys(mode.states).reduce((acc, state) => {
        acc[state] = update;
        return acc;
      }, {} as Record<string, string>);
    } else if ((isModeValue(originalValue) || originalValue === undefined) && isModeValue(update)) {
      return {
        ...originalValue,
        ...update,
      };
    } else if (isGlobal) {
      return update;
    } else {
      console.warn('The value for this token cannot be merged into the theme:', token);
    }
  }

  // Merge root-level tokens into the theme
  entries(override.tokens).forEach(([token, update]) => {
    const newValue = withTokenApplied(theme.tokens[token], token, update);
    if (newValue) {
      result.tokens[token] = newValue;
    }
  });

  // Merge context-specific tokens into each context
  if (override.contexts) {
    entries(override.contexts).forEach(([contextId, context]) => {
      const resultContext = result.contexts[contextId];

      if (!context || !resultContext) {
        return;
      }

      entries(context.tokens).forEach(([token, update]) => {
        const originalValue = resultContext.tokens[token] ?? result.tokens[token];
        const newValue = withTokenApplied(originalValue, token, update);
        if (newValue) {
          result.contexts[contextId].tokens[token] = newValue;
        }
      });
    });
  }
  return result;
}
