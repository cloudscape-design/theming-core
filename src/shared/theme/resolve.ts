// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { Context, Mode } from '.';
import { cloneDeep } from '../utils';
import { Theme, Value } from './interfaces';
import { getDefaultState, getMode, getReference, isModeValue, isReference } from './utils';

export type ModeTokenResolution = Record<string, Value>;
export type SpecificTokenResolution = Value;
interface ThemeResolution<TR> {
  [token: string]: TR;
}

type ModeTokenResolutionPath = Record<string, Array<string>>;
type SpecificTokenResolutionPath = Array<string>;

export type FullResolution = ThemeResolution<ModeTokenResolution | SpecificTokenResolution>;
export type SpecificResolution = ThemeResolution<SpecificTokenResolution>;
export type FullResolutionPaths = ThemeResolution<ModeTokenResolutionPath | SpecificTokenResolutionPath>;

interface FullResolutionWithPaths {
  resolvedTheme: FullResolution;
  resolutionPaths: FullResolutionPaths;
}

export function resolveTheme(theme: Theme): FullResolution {
  return resolveThemeWithPaths(theme).resolvedTheme;
}
export function resolveThemeWithPaths(theme: Theme): FullResolutionWithPaths {
  const resolvedTheme: FullResolution = {};
  const resolutionPaths: FullResolutionPaths = {};

  Object.keys(theme.tokens).forEach((token) => {
    const mode = getMode(theme, token);
    if (mode) {
      resolutionPaths[token] = {};
      resolvedTheme[token] = Object.keys(mode.states).reduce((acc, state: string) => {
        (resolutionPaths[token] as ModeTokenResolutionPath)[state] = [];
        acc[state] = resolveToken(theme, token, (resolutionPaths[token] as ModeTokenResolutionPath)[state], state);
        return acc;
      }, {} as Record<string, string>);
    } else {
      resolutionPaths[token] = [];
      resolvedTheme[token] = resolveToken(theme, token, resolutionPaths[token] as SpecificTokenResolutionPath);
    }
  });

  return { resolvedTheme, resolutionPaths };
}

function resolveToken(theme: Theme, token: string, path: Array<string>, state?: string): string {
  if (!theme.tokens[token]) {
    throw new Error(`Token ${token} does not exist in the theme.`);
  }
  if (path.indexOf(token) !== -1) {
    throw new Error(`Token ${token} has a circular dependency.`);
  }
  path.push(token);
  let assignment = theme.tokens[token];

  if (isModeValue(assignment)) {
    if (!state) {
      throw new Error('Mode resolution needs state');
    }
    assignment = assignment[state];
  }

  if (isReference(assignment)) {
    const ref = getReference(assignment);
    return resolveToken(theme, ref, path, state);
  } else {
    return assignment;
  }
}

export function resolveContext(theme: Theme, context: Context): FullResolution {
  const tmp = cloneDeep(theme);
  tmp.tokens = {
    ...tmp.tokens,
    ...context.tokens,
  };
  return resolveTheme(tmp);
}

type Reducer = (
  tokenResolution: ModeTokenResolution | SpecificTokenResolution,
  token: string,
  theme: Theme
) => SpecificTokenResolution | undefined;

export function reduce(
  resolution: FullResolution | SpecificResolution,
  theme: Theme,
  reducer: Reducer
): SpecificResolution {
  return Object.keys(resolution).reduce((acc, token) => {
    const reduced = reducer(resolution[token], token, theme);
    if (reduced) {
      acc[token] = reduced;
    }
    return acc;
  }, {} as SpecificResolution);
}

export const defaultsReducer =
  () => (tokenResolution: ModeTokenResolution | SpecificTokenResolution, token: string, theme: Theme) => {
    const mode = getMode(theme, token);
    if (mode && isModeTokenResolution(tokenResolution)) {
      const defaultState = getDefaultState(mode);
      return tokenResolution[defaultState];
    } else if (isSpecificTokenResolution(tokenResolution)) {
      return tokenResolution;
    } else {
      throw new Error(`Mismatch between resolution ${tokenResolution} and mode ${mode}`);
    }
  };

export const modeReducer =
  (mode: Mode, state: string) =>
  (tokenResolution: ModeTokenResolution | SpecificTokenResolution, token: string, theme: Theme) => {
    const tokenMode = getMode(theme, token);
    if (tokenMode && tokenMode.id === mode.id && isModeTokenResolution(tokenResolution)) {
      return tokenResolution[state];
    } else if (isSpecificTokenResolution(tokenResolution)) {
      return tokenResolution;
    }
  };

export function difference(base: SpecificResolution, other: SpecificResolution): SpecificResolution;
export function difference(base: FullResolution, other: FullResolution): FullResolution;
export function difference(
  base: FullResolution | SpecificResolution,
  other: FullResolution | SpecificResolution
): FullResolution {
  const result: FullResolution = {};

  Object.keys(other).forEach((token) => {
    const baseVal = base[token];
    const otherVal = other[token];

    if (isSpecificTokenResolution(baseVal) && isSpecificTokenResolution(otherVal) && baseVal !== otherVal) {
      result[token] = otherVal;
    } else if (isModeTokenResolution(baseVal) && isModeTokenResolution(otherVal)) {
      const resolved = Object.keys(otherVal).reduce((acc, state) => {
        if (baseVal[state] !== otherVal[state]) {
          acc[state] = otherVal[state];
        }
        return acc;
      }, {} as ModeTokenResolution);
      if (!isEmpty(resolved)) {
        result[token] = resolved;
      }
    }
  });

  return result;
}
export function isModeTokenResolution(val: SpecificTokenResolution | ModeTokenResolution): val is ModeTokenResolution {
  return typeof val === 'object';
}

export function isSpecificTokenResolution(
  val: SpecificTokenResolution | ModeTokenResolution
): val is SpecificTokenResolution {
  return typeof val === 'string';
}

const isEmpty = (obj: Record<string, unknown>) => Object.keys(obj).length === 0;
