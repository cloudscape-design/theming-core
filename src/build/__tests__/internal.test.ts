// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { join } from 'node:path';
import fs from 'node:fs';
import { test, expect } from 'vitest';

import { outputDir, preset, scssDir, presetPath } from './common';
import { buildThemedComponentsInternal, BuildThemedComponentsInternalParams } from '../internal';
import { ThemePreset } from '../../shared/theme';

const presetToParams = (preset: ThemePreset) => ({
  primary: preset.theme,
  exposed: preset.exposed,
  themeable: preset.themeable,
  variablesMap: preset.variablesMap,
});

test('builds internal themed components without errors', async () => {
  const internalOutputDir = join(outputDir, 'internal-test');
  const componentsOutputDir = join(internalOutputDir, 'components');
  const designTokensOutputDir = join(internalOutputDir, 'design-tokens');
  const params: BuildThemedComponentsInternalParams = {
    ...presetToParams(preset),
    componentsOutputDir,
    designTokensOutputDir,
    scssDir,
  };
  await buildThemedComponentsInternal(params);
  expect(
    fs.readFileSync(join(componentsOutputDir, 'internal/base-component/styles.scoped.css'), 'utf-8')
  ).toMatchSnapshot();
});

test('throws error if designTokensOutputDir not specified', async () => {
  const internalOutputDir = join(outputDir, 'error-test');
  const componentsOutputDir = join(internalOutputDir, 'components');
  const params = {
    ...presetToParams(preset),
    componentsOutputDir,
    scssDir,
  };
  return expect(() => buildThemedComponentsInternal(params)).rejects.toThrow();
});

test('skips design tokens and preset if specified', async () => {
  const internalOutputDir = join(outputDir, 'skipped-test');
  const componentsOutputDir = join(internalOutputDir, 'components');
  const params: BuildThemedComponentsInternalParams = {
    ...presetToParams(preset),
    componentsOutputDir,
    scssDir,
    skip: ['design-tokens', 'preset'],
  };

  await buildThemedComponentsInternal(params);

  expect(fs.readdirSync(internalOutputDir)).toHaveLength(1);
  expect(fs.existsSync(presetPath(componentsOutputDir))).toBeFalsy();
});
