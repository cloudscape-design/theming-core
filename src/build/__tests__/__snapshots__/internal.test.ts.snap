// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`builds internal themed components without errors 1`] = `
":root {
  --font-family-base-91xi75:"Amazon Ember", "Helvetica Neue", Roboto, Arial, sans-serif;
  --color-background-button-primary-default-nhput1:#ec7211;
  --color-background-button-primary-active-xbf2p5:#dd6b10;
  --space-scaled-xs-yqelf2:8px;
}

.compact-mode:not(#\\9) {
  --space-scaled-xs-yqelf2:4px;
}

.dark-mode:not(#\\9) {
  --color-background-button-primary-active-xbf2p5:#ec7211;
}"
`;
